File: app/core/geo_utils.py
"""Geographic utility functions for Michi-Navi.

Provides tile coordinate conversion, distance/bearing calculations,
and direction filtering for the map engine and POI search.
"""

import math

# Earth radius in km (mean)
_EARTH_RADIUS_KM = 6371.0


def deg2tile(lat_deg: float, lon_deg: float, zoom: int) -> tuple[int, int]:
    """Convert latitude/longitude to OSM slippy-map tile coordinates.

    Args:
        lat_deg: Latitude in degrees.
        lon_deg: Longitude in degrees.
        zoom: Zoom level (0-18).

    Returns:
        (x, y) tile coordinates.
    """
    lat_rad = math.radians(lat_deg)
    n = 2.0 ** zoom
    x = int((lon_deg + 180.0) / 360.0 * n)
    y = int((1.0 - math.asinh(math.tan(lat_rad)) / math.pi) / 2.0 * n)
    return x, y


def tile2deg(x: int, y: int, zoom: int) -> tuple[float, float]:
    """Convert tile coordinates to the north-west corner lat/lon.

    Args:
        x: Tile x coordinate.
        y: Tile y coordinate.
        zoom: Zoom level (0-18).

    Returns:
        (lat_deg, lon_deg) of the tile's north-west corner.
    """
    n = 2.0 ** zoom
    lon_deg = x / n * 360.0 - 180.0
    lat_rad = math.atan(math.sinh(math.pi * (1 - 2 * y / n)))
    lat_deg = math.degrees(lat_rad)
    return lat_deg, lon_deg


def haversine(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate the great-circle distance between two points (km).

    Uses the Haversine formula with a spherical Earth model.

    Args:
        lat1, lon1: First point in degrees.
        lat2, lon2: Second point in degrees.

    Returns:
        Distance in kilometres.
    """
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(dlon / 2) ** 2
    )
    return _EARTH_RADIUS_KM * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def bearing(lat1: float, lon1: float, lat2: float, lon2: float) -> float:
    """Calculate the initial bearing from point 1 to point 2.

    Args:
        lat1, lon1: Origin point in degrees.
        lat2, lon2: Destination point in degrees.

    Returns:
        Bearing in degrees (0-360, clockwise from north).
    """
    dlon = math.radians(lon2 - lon1)
    x = math.sin(dlon) * math.cos(math.radians(lat2))
    y = (
        math.cos(math.radians(lat1)) * math.sin(math.radians(lat2))
        - math.sin(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.cos(dlon)
    )
    return (math.degrees(math.atan2(x, y)) + 360) % 360


def is_ahead(course: float, bearing_to_poi: float, threshold: float = 45) -> bool:
    """Check whether a POI is within the forward cone of travel.

    Args:
        course: Current heading in degrees (0-360).
        bearing_to_poi: Bearing from current position to the POI (0-360).
        threshold: Half-angle of the forward cone in degrees.

    Returns:
        True if the POI is within ±threshold of the course.
    """
    diff = abs(course - bearing_to_poi)
    if diff > 180:
        diff = 360 - diff
    return diff <= threshold

File: MichiNavi/Shared/Services/GeoUtils.swift
import Foundation
import CoreLocation

/// 地理計算ユーティリティ（raspberry-pi 版 geo_utils.py の Swift 移植）
enum GeoUtils {

    private static let earthRadiusKm: Double = 6371.0

    /// Haversine 公式による2点間の大圏距離 (km)
    static func haversine(
        from: CLLocationCoordinate2D,
        to: CLLocationCoordinate2D
    ) -> Double {
        let lat1 = from.latitude.radians
        let lat2 = to.latitude.radians
        let dLat = (to.latitude - from.latitude).radians
        let dLon = (to.longitude - from.longitude).radians

        let a = sin(dLat / 2) * sin(dLat / 2)
            + cos(lat1) * cos(lat2) * sin(dLon / 2) * sin(dLon / 2)
        let c = 2 * atan2(sqrt(a), sqrt(1 - a))
        return earthRadiusKm * c
    }

    /// 2点間の初期方位角 (0–360°、北=0、時計回り)
    static func bearing(
        from: CLLocationCoordinate2D,
        to: CLLocationCoordinate2D
    ) -> Double {
        let lat1 = from.latitude.radians
        let lat2 = to.latitude.radians
        let dLon = (to.longitude - from.longitude).radians

        let x = sin(dLon) * cos(lat2)
        let y = cos(lat1) * sin(lat2) - sin(lat1) * cos(lat2) * cos(dLon)
        let rad = atan2(x, y)
        return (rad.degrees + 360).truncatingRemainder(dividingBy: 360)
    }

    /// 進行方向の前方コーン内にあるか判定
    /// - Parameters:
    ///   - heading: 車両の進行方向 (0–360°)
    ///   - bearingToTarget: 目標への方位角 (0–360°)
    ///   - threshold: 前方コーンの半角 (デフォルト 45°)
    static func isAhead(
        heading: Double,
        bearingToTarget: Double,
        threshold: Double = 45
    ) -> Bool {
        var diff = bearingToTarget - heading
        // -180〜180 に正規化
        if diff > 180 { diff -= 360 }
        if diff < -180 { diff += 360 }
        return abs(diff) <= threshold
    }

    /// 方位角から 16 方位の文字列に変換
    static func cardinalDirection(from bearing: Double) -> String {
        let cardinals = [
            "N", "NNE", "NE", "ENE",
            "E", "ESE", "SE", "SSE",
            "S", "SSW", "SW", "WSW",
            "W", "WNW", "NW", "NNW"
        ]
        let index = Int((bearing + 11.25).truncatingRemainder(dividingBy: 360) / 22.5)
        return cardinals[index % 16]
    }
}

// MARK: - Double 拡張

private extension Double {
    var radians: Double { self * .pi / 180 }
    var degrees: Double { self * 180 / .pi }
}
